server/src/main/java/org/opensearch/search/startree/StarTreeQueryHelper.java

```diff
@@ -0,0 +1,161 @@
+        SegmentReader reader = Lucene.segmentReader(context.reader());
+        if (!(reader.getDocValuesReader() instanceof CompositeIndexReader)) {
+            return null;
+        }
+        CompositeIndexReader starTreeDocValuesReader = (CompositeIndexReader) reader.getDocValuesReader();
+        return (StarTreeValues) starTreeDocValuesReader.getCompositeIndexValues(starTree);
+    }
+
+    /**
+     * Get the star-tree leaf collector
+     * This collector computes the aggregation prematurely and invokes an early termination collector
+     */
+    public static LeafBucketCollector getStarTreeLeafCollector(
+        SearchContext context,
+        ValuesSource.Numeric valuesSource,
+        LeafReaderContext ctx,
+        LeafBucketCollector sub,
+        CompositeIndexFieldInfo starTree,
+        String metric,
+        Consumer<Long> valueConsumer,
+        Runnable finalConsumer
+    ) throws IOException {
+        StarTreeValues starTreeValues = getStarTreeValues(ctx, starTree);
+        assert starTreeValues != null;
+        String fieldName = ((ValuesSource.Numeric.FieldData) valuesSource).getIndexFieldName();
+        String metricName = StarTreeUtils.fullyQualifiedFieldNameForStarTreeMetricsDocValues(starTree.getField(), fieldName, metric);
+
+        assert starTreeValues != null;
+        SortedNumericStarTreeValuesIterator valuesIterator = (SortedNumericStarTreeValuesIterator) starTreeValues.getMetricValuesIterator(
+            metricName
+        );
+        // Obtain a FixedBitSet of matched star tree document IDs
+        FixedBitSet filteredValues = getStarTreeFilteredValues(context, ctx, starTreeValues);
+        assert filteredValues != null;
+
+        int numBits = filteredValues.length();  // Get the number of the filtered values (matching docs)
+        if (numBits > 0) {
+            // Iterate over the filtered values
+            for (int bit = filteredValues.nextSetBit(0); bit != DocIdSetIterator.NO_MORE_DOCS; bit = (bit + 1 < numBits)
+                ? filteredValues.nextSetBit(bit + 1)
+                : DocIdSetIterator.NO_MORE_DOCS) {
+                // Advance to the entryId in the valuesIterator
+                if (valuesIterator.advanceExact(bit) == false) {
+                    continue;  // Skip if no more entries
+                }
+
+                // Iterate over the values for the current entryId
+                for (int i = 0, count = valuesIterator.entryValueCount(); i < count; i++) {
+                    long value = valuesIterator.nextValue();
+                    valueConsumer.accept(value); // Apply the consumer operation (e.g., max, sum)
+                }
+            }
+        }
+
+        // Call the final consumer after processing all entries
+        finalConsumer.run();
+
+        // Return a LeafBucketCollector that terminates collection
+        return new LeafBucketCollectorBase(sub, valuesSource.doubleValues(ctx)) {
```

Comment:
I think either in Sandesh's PR or this PR - we should stop using leafBuckerCollector - this approach as we debugged yesterday also executes the original query scorer which increases the query latency for all star tree queries.

We should follow the new interfaces used in date histogram - where getLeafCollector throws collectionTerminatedException

cc: @sandeshkr419

Reply:
@sandeshkr419 is handling the same.